features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/transport/main.cpp

```diff
@@ -0,0 +1,190 @@
+
+using ble::vendor::cordio::CordioHCIHook;
+
+#define RESET_COMMAND_TIMEOUT (10 * 1000)
+#define RESET_PARAMETER_LENGTH  4
+#define RESET_EXPECTED_STATUS  0
+#define HCI_OPCODE_RESET_LSB (HCI_OPCODE_RESET & 0xFF)
+#define HCI_OPCODE_RESET_MSB (HCI_OPCODE_RESET >> 8)
+
+enum test_result_t {
+    TEST_RESULT_TIMEOUT_FAILURE,
+    TEST_RESULT_FAILURE,
+    TEST_RESULT_SUCCESS
+};
+
+enum state_t {
+    WAITING_EVENT_PACKET,
+    WAITING_EVENT_CODE_COMPLETE,
+    WAITING_PARAMETER_LENGTH,
+    WAITING_STATUS,
+    WAITING_NUM_HCI_EVT_PACKET,
+    WAITING_OPCODE_LSB,
+    WAITING_OPCODE_MSB,
+    DONE
+};
+
+static state_t state = WAITING_EVENT_PACKET;
+static test_result_t test_result = TEST_RESULT_TIMEOUT_FAILURE;
+
+static rtos::Semaphore sem;
+
+static uint8_t reset_cmd[] = {
+    HCI_OPCODE_RESET_LSB, HCI_OPCODE_RESET_MSB,  // reset opcode
+    0 // parameter length
+};
+
+static void hci_driver_rx_dummy_handler(uint8_t* data, uint8_t len) { }
+
+static void hci_driver_rx_reset_handler(uint8_t* data, uint8_t len) {
+    for (size_t i = 0; i < len; ++i) {
+        switch (state) {
+            case WAITING_EVENT_PACKET:
+                if (data[i] == HCI_EVT_TYPE) {
+                    state = WAITING_EVENT_CODE_COMPLETE;
+                } else {
+                    test_result = TEST_RESULT_FAILURE;
+                }
+                break;
+
+            case WAITING_EVENT_CODE_COMPLETE:
+                if (data[i] == HCI_CMD_CMPL_EVT) {
+                    state = WAITING_PARAMETER_LENGTH;
+                } else {
+                    test_result = TEST_RESULT_FAILURE;
+                }
+                break;
+
+            case WAITING_PARAMETER_LENGTH:
+                if (data[i] == RESET_PARAMETER_LENGTH) {
+                    state = WAITING_NUM_HCI_EVT_PACKET;
+                } else {
+                    test_result = TEST_RESULT_FAILURE;
+                }
+                break;
+
+            case WAITING_NUM_HCI_EVT_PACKET:
+                // controler dependent; can be any value, pass on to the next token
+                state = WAITING_OPCODE_LSB;
+                break;
+
+            case WAITING_OPCODE_LSB:
+                if (data[i] == HCI_OPCODE_RESET_LSB) {
+                    state = WAITING_OPCODE_MSB;
+                } else {
+                    test_result = TEST_RESULT_FAILURE;
+                }
+                break;
+
+            case WAITING_OPCODE_MSB:
+                if (data[i] == HCI_OPCODE_RESET_MSB) {
+                    state = WAITING_STATUS;
+                } else {
+                    test_result = TEST_RESULT_FAILURE;
+                }
+                break;
+
+            case WAITING_STATUS:
+                if (data[i] == RESET_EXPECTED_STATUS) {
+                    test_result = TEST_RESULT_SUCCESS;
+                    state = DONE;
+                } else {
+                    test_result = TEST_RESULT_FAILURE;
+                }
+                break;
+        }
+
```

Comment:
You should break from the for loop here if `test_result` is `TEST_RESULT_FAILURE`

Reply:
doesn't the if do that?

Reply:
Yes, the if does that whenever a result is present (success or failure).

Reply:
OK - there are a few things that confused me, now I understand what you're doing. Few questions/remarks:

- Could you rename `TEST_RESULT_TIMEOUT_FAILURE` to something more explicit (like 'in progress'); it hasn't timed out till proven otherwise
- At the moment, if you don't know the HCI OpCode's value, you skip to the next byte - which would be parameters' length; what's the rationale behind this? You could interpret the length field (and subsequent parameters) as OpCodes in the current implementation